contracts/chain-adapters/ForwarderBase.sol

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.0;
+
+import { UUPSUpgradeable } from "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+
+/**
+ * @title ForwarderBase
+ * @notice This contract expects to receive messages and tokens from the Hub Pool on L1. It rejects messages which do not originate
+ * from the hub pool.
+ * @notice Base contract deployed some network between L1 and the network housing a spoke pool contract. If a message is sent
+ * to this contract which came from the cross domain admin, then it is routed to the next network via the canonical messaging
+ * bridge (which may be done automatically by the network's finalization infrastructure). Tokens sent from L1 are stored temporarily
+ * on this contract. Any EOA can initiate a bridge of these tokens to the target `l3SpokePool`.
+ * @custom:security-contact bugs@across.to
+ */
+abstract contract ForwarderBase is UUPSUpgradeable {
+    // L3 address of the recipient of L1 messages and tokens.
+    address public l3SpokePool;
+
+    // L1 address of the contract which can relay messages to the l3SpokePool contract and update this proxy contract.
+    address public crossDomainAdmin;
+
+    event TokensForwarded(address indexed l2Token, uint256 amount);
+    event MessageForwarded(address indexed target, bytes message);
+    event SetXDomainAdmin(address indexed crossDomainAdmin);
+    event SetL3SpokePool(address indexed l3SpokePool);
+
+    error InvalidCrossDomainAdmin();
+    error InvalidL3SpokePool();
+
+    /*
+     * @dev All functions with this modifier must revert if msg.sender != crossDomainAdmin. Each L2 may have
+     * unique aliasing logic, so it is up to the chain-specific forwarder contract to verify that the sender is valid.
+     */
+    modifier onlyAdmin() {
+        _requireAdminSender();
+        _;
+    }
+
+    /**
+     @notice Constructs the Forwarder contract.
+     @dev _disableInitializers() restricts anybody from initializing the implementation contract, which if not done, 
+     * may disrupt the proxy if another EOA were to initialize it.
+     */
+    constructor() {
+        _disableInitializers();
+    }
+
+    /**
+     * @notice Initializes the forwarder contract.
+     * @param _l3SpokePool L3 address of the contract which will receive messages and tokens which are temporarily
+     * stored in this contract on L2.
+     * @param _crossDomainAdmin L1 address of the contract which can send root bundles/messages to this forwarder contract.
+     * In practice, this is the hub pool.
+     */
+    function __Forwarder_init(address _l3SpokePool, address _crossDomainAdmin) public onlyInitializing {
+        __UUPSUpgradeable_init();
```

Comment:
Should this contract be upgradeable? I think so as it could be helpful to re-deploy updates and not have to update the RerouterAdapter to point to new forwarder contracts, but I want to make sure we've carefully considered why we are making this upgradeable. I assume you made it upgradeable by default because you copied the SpokePool's code.

Reply:
https://github.com/across-protocol/contracts/pull/609#discussion_r1763825724